Basic name poisoning support

bricknerb opened the pull request:
https://github.com/carbon-language/carbon-lang/issues/4622
When using an unqualified name, disallow declaring that name in all scopes that would make it ambiguous in retrospect.
Doesn't include support for poisoning in `impl library` (see new test for that with TODO).
Implemented by introduce `InstId::PoisonedName` and entries with it to `NameScope`.

bricknerb committed:
Basic name poisoning support
https://github.com/carbon-language/carbon-lang/issues/4622
When using an unqualified name disallow using that name in all scopes that would make it ambiguous in retrospect.
Doesn't include support for poisoning when importing (see new test for that with TODO).
Implemented by introduce `InstId::PoisonedName` and entries with it to `NameScope`.

jonmeow commented on toolchain/check/testdata/function/declaration/no_prelude/name_poisoning.carbon:
```diff
@@ -12,48 +12,152 @@
+// CHECK:STDERR: fail_poison_import: error: file extension of `.carbon` required for api [IncorrectExtension]
+// CHECK:STDERR:
+alias N.C = C;
+
```
Is a name reference to the poisoned entity tested? e.g.:

```suggestion

// --- fail_use_poison.carbon

library "[[@TEST_NAME]]";

class C {}

namespace N;
fn N.F() -> C;

fn F() -> N.C;

```

bricknerb replied:
Done, though this was implicitly tested in fail_poison_with_usage.

jonmeow commented on toolchain/check/testdata/function/declaration/no_prelude/name_poisoning.carbon:
```diff
@@ -12,48 +12,152 @@
+// CHECK:STDERR:
+alias N.C = C;
+
+// --- fail_poison_import
```
```suggestion
// --- fail_poison_import.carbon
```

bricknerb replied:
Done.

jonmeow commented on toolchain/check/testdata/function/declaration/no_prelude/name_poisoning.carbon:
```diff
@@ -12,48 +12,152 @@
+
+// --- fail_poison_import
+
+import library "poison";
```
```suggestion
library "[[@TEST_NAME]]";
import library "poison";
```

Should declare the library name

bricknerb replied:
Done.

jonmeow commented on toolchain/sem_ir/name_scope.h:
```diff
@@ -51,6 +51,7 @@ class NameScope : public Printable<NameScope> {
 
   // Searches for the given name and returns an EntryId if found or nullopt if
   // not.
+  // If the name is found, it might be a poisoned name.
```
Generally I'd suggest writing comments like you would markdown:

- Put a blank line inbetween if you want separate paragraphs. i.e.:

```suggestion
  // not.
  //
  // If the name is found, it might be a poisoned name.
```

- Wrap if you want a single paragraph:

```suggestion
  // not. If the name is found, it might be a poisoned name.
```

- Use bullets (`// - text here`) if you're trying to list things. (probably doesn't apply in this comment)

Just noting this for style of below comments too.

bricknerb replied:
Done.

jonmeow commented on toolchain/sem_ir/name_scope.cpp:
```diff
@@ -46,9 +52,25 @@ auto NameScope::LookupOrAdd(SemIR::NameId name_id, InstId inst_id,
   }
 
+  CARBON_CHECK(!inst_id.is_poisoned(),
+               "Cannot add a poisoned name: {0}. Use AddPoison()", name_id);
```
Should this check before the insert, so that it fails even if the name was previously inserted?

bricknerb replied:
Done.

jonmeow commented on toolchain/sem_ir/name_scope.cpp:
```diff
@@ -46,9 +52,25 @@ auto NameScope::LookupOrAdd(SemIR::NameId name_id, InstId inst_id,
+  }
+
+  CARBON_CHECK(GetEntry(insert_result.value()).inst_id.is_poisoned(),
+               "Trying to poison an existing non poisoned name: {0}", name_id);
```
```suggestion
               "Trying to poison an existing non-poisoned name: {0}", name_id);
```

bricknerb replied:
Done.

jonmeow commented on toolchain/check/context.cpp:
```diff
@@ -326,16 +334,26 @@ auto Context::LookupUnqualifiedName(Parse::NodeId node_id,
 
   // Walk the non-lexical scopes and perform lookups into each of them.
+  // Collect scopes to poison this name when it's found.
+  std::vector<LookupScope> scopes_to_poison;
```
We prefer LLVM types over std types in most cases.

```suggestion
  llvm::SmallVector<LookupScope> scopes_to_poison;
```

bricknerb replied:
Done.

jonmeow commented on toolchain/check/context.cpp:
```diff
@@ -326,16 +334,26 @@ auto Context::LookupUnqualifiedName(Parse::NodeId node_id,
+
       return non_lexical_result;
     }
+    scopes_to_poison.push_back(
```
Had you considered not saving scopes where `non_lexical_result.inst_id.is_poisoned()` is true, in order to avoid poisoning twice? Would that allow `AddPoison` to error on _all_ insert failures, not just non-poisoned ones?

bricknerb replied:
Done.

jonmeow commented on toolchain/sem_ir/ids.h:
```diff
@@ -39,12 +39,23 @@ struct InstId : public IdBase<InstId> {
   using IdBase::IdBase;
 
+  constexpr auto is_poisoned() const -> bool {
+    return index == PoisonedNameIndex;
```
```suggestion
    return *this == PoisonedName;
```

bricknerb replied:
Done.

jonmeow commented on toolchain/sem_ir/ids.h:
```diff
@@ -39,12 +39,23 @@ struct InstId : public IdBase<InstId> {
 };
 
 constexpr InstId InstId::Invalid = InstId(InvalidIndex);
+constexpr InstId InstId::PoisonedName = InstId(PoisonedNameIndex);
```
```suggestion
constexpr InstId InstId::PoisonedName = InstId(InvalidIndex - 1);
```

bricknerb replied:
Done.

jonmeow commented on toolchain/sem_ir/ids.h:
```diff
@@ -39,12 +39,23 @@ struct InstId : public IdBase<InstId> {
+  }
+
   auto Print(llvm::raw_ostream& out) const -> void;
+
+  static constexpr int32_t PoisonedNameIndex = InvalidIndex - 1;
```
```suggestion
```

For style with [declaration order](https://google.github.io/styleguide/cppguide.html#Declaration_Order), this static constant would belong at the top. However, I'm suggesting a few changes to remove this, similar to what we have elsewhere in this file.

bricknerb replied:
Done.

jonmeow commented on toolchain/check/decl_name_stack.cpp:
```diff
@@ -276,9 +285,15 @@ static auto CheckQualifierIsResolved(
       CARBON_FATAL("No qualifier to resolve");
 
     case DeclNameStack::NameContext::State::Resolved:
+      if (name_context.resolved_inst_id.is_poisoned()) {
+        context.DiagnoseNameNotFound(name_context.loc_id,
+                                     name_context.unresolved_name_id);
+        return false;
+      }
       return true;
 
     case DeclNameStack::NameContext::State::Unresolved:
+    case DeclNameStack::NameContext::State::Poisoned:
```
```suggestion
    case DeclNameStack::NameContext::State::Poisoned:
    case DeclNameStack::NameContext::State::Unresolved:
```

Very minor, but when there's no particular significance to the ordering, I'll suggest keeping things in lexical order.

bricknerb replied:
Done.

jonmeow commented on toolchain/check/handle_class.cpp:
```diff
@@ -113,6 +113,12 @@ static auto MergeOrAddName(Context& context, Parse::AnyClassDeclId node_id,
   }
 
+  if (prev_id.is_poisoned()) {
+    // This is declaration of a poisoned name.
```
```suggestion
    // This is a declaration of a poisoned name.
```

bricknerb replied:
Done.

jonmeow commented on toolchain/check/context.cpp:
```diff
@@ -216,6 +216,14 @@ auto Context::DiagnoseDuplicateName(SemIRLoc dup_def, SemIRLoc prev_def)
+      NameDeclPoisoned, Error,
+      "cannot declare this name in this scope since is was already used "
+      "without qualification in the context of this scope");
+  emitter_->Build(dup_def, NameDeclPoisoned).Emit();
```
From this diagnostic, I'm thinking that we should take a different approach that allows us to annotate the location which introduced the poisoned name. i.e., something like:

```
// error: the NameDeclPoisoned error
// class N.C;
//       ^~~
// note: unqualified name lookup here
// fn N.F() -> C;
//             ^
```

That might require changing the way poisoned entries are tracked, to keep the declaration that _caused_ the poisoning with the `NameScope::Entry`. This is having me wonder if we should have something like:

```
  struct Entry {
    NameId name_id;
    // Maybe this should actually be a `union` with a `LocId` based on `is_poisoned`?
    InstId inst_id;
    AccessKind access_kind;
    bool is_poisoned;
  };
  static_assert(sizeof(Entry) == 12);
```

(noting the `static_assert` to emphasize packing)

The reason I'm thinking in that direction is because I'm now thinking we should actually use `inst_id` here to point back to the instruction which introduced the poisoning (there may be a better way of doing that, too).

Sorry, I know I'd commented in favor of the special `InstId`. I think though that this could be done as an incremental change on top of what you have here. I'd be fine if you want to leave something like this as a TODO or for a separate PR. I just wanted to note the thought that we may need to change approaches for a better diagnostic. Let me know how you'd like to proceed here.

bricknerb replied:
Thanks Jon!
Yes, this was my concern with this option in https://docs.google.com/document/d/16vhcVeZN2w8iIl_a6gNxd5kFZPPfWh3f5_TnaK9sdHQ

> Since it uses the existing data structure, it’s somewhat complex to add specific metadata for diagnosis. For example, adding a reference to where the name was poisoned doesn’t fit well into this design.

I thought of manipulating `Entry` for that but there is some concern that it makes the structure more complex to use and extend, compared to having a separate set, which might make the lookup more complex.
Since this might change this PR significantly, I think I'll make that a TODO so we can have name poisoning support and improve the diagnostics in a separate PR as suggested.
Added a TODO in `Context::DiagnosePoisonedName()`.

jonmeow replied:
I think we're likely to prefer the structure complexity versus a new set because it's likely higher performance.

jonmeow commented on toolchain/check/testdata/function/declaration/no_prelude/name_poisoning.carbon:
```diff
@@ -12,48 +12,152 @@
+
 // TODO: Should fail here since C was poisoned for namespace N when it was used
-// in N context without qualification.
+//       in N context without qualification in the imported library.
```
As discussed, this shouldn't fail because it would be cross-library. However, it should fail for an `impl library`, such as:

```
// --- fail_poison.impl.carbon

// "TEST_NAME" here translates to "poison", which is what you'd want.
impl library "[[@TEST_NAME]]";

class N.C {}
```

bricknerb replied:
Done.

jonmeow commented on toolchain/check/context.cpp:
```diff
@@ -216,6 +216,14 @@ auto Context::DiagnoseDuplicateName(SemIRLoc dup_def, SemIRLoc prev_def)
+auto Context::DiagnosePoisonedName(SemIRLoc dup_def) -> void {
+  CARBON_DIAGNOSTIC(
+      NameDeclPoisoned, Error,
+      "cannot declare this name in this scope since is was already used "
+      "without qualification in the context of this scope");
```
```suggestion
      "name previously used by unqualified name lookup and not found; cannot later be declared");
```

jonmeow replied:
We have a little [diagnostic style guide](https://github.com/carbon-language/carbon-lang/blob/trunk/toolchain/docs/diagnostics.md#diagnostic-message-style-guide). Generally I think we're trying to focus on the situation the toolchain observed first. Suggesting a rephrasing which I _think_ better reflects the intent.

zygoloid replied:
I wonder if it'd be better to put the diagnostic at the point of use, and say something like "name used before it was declared" with a note pointing to the later declaration?

jonmeow replied:
That seems reasonable, though noting [my comments about location above](https://github.com/carbon-language/carbon-lang/pull/4654#discussion_r1876530039), so for now this would probably look like using an `Invalid` ID for the location that needs to be piped through:

```
CARBON_DIAGNOSTIC(NameUseBeforeDecl, Error, "name used before it was declared");
CARBON_DIAGNOSTIC(NameUseBeforeDeclNote, Note, "declared here");
// TODO: This should pass in the original declaration.
unit_info.emitter.Emit(Parse::NodeId::Invalid, NameUseBeforeDecl)
    .Note(decl_loc, NameUseBeforeDeclNote)
    .Emit();
```

bricknerb replied:
Thanks!
I've changed the diagnostic name and text.
It makes sense to put the warning where the name is poisoned, though I find an error in code that is caused because of code that comes later somewhat confusing.
I'm ok with both options.
Anyways, this needs to wait for when we store the data on what caused the poisoning.

jonmeow replied:
Unresolving -- I think we should probably go with the phrasing zygoloid suggested, with TODOs per https://github.com/carbon-language/carbon-lang/pull/4654#discussion_r1876772138

bricknerb replied:
Right now we have the location of the later declaration.
IIUC, the suggestion is to error where the poisoning happened with a note on the later declaration.
If we go with this suggestion, we only have the note, not the error.
Can you clarify what you suggest for this PR?
I've made the TODO clearer.

jonmeow replied:
Sure, I've copied in my code from the above comment into a suggested edit: https://github.com/carbon-language/carbon-lang/pull/4654#discussion_r1880640674.

jonmeow replied:
Can you elaborate why you think there will only be the note and not the error? What does the delta to tests look like?

bricknerb replied:
Thanks for the clarification.
Done.
See my reply in the other comment.

jonmeow reviewed (commented):
Looks good! This made me think a little more about what's needed to get a good diagnostic -- some comments inline, sorry I didn't think about that before.

danakj committed:
Disable modernize-use-trailing-return-type on MATCHER_P (#4656)

clang-tidy gives a false positive on the use of the MATCHER_P macro.

zygoloid committed:
Improve diagnostics for missing qualified names. (#4638)

Mention the scope in which the name wasn't found.

---------

Co-authored-by: Geoff Romer <gromer@google.com>

zygoloid committed:
Track complete types required by a generic. (#4652)

When a generic requires a symbolic type to be complete, add a new
`require_complete_type` instruction to the generic eval block. During
monomorphization of such an instruction, require that type to be
complete.

bricknerb committed:
Address some review comments:
* Use `llvm::SmallVector` instead of `std::vector`.
* Put `DeclNameStack::NameContext::State::Poisoned` before `DeclNameStack::NameContext::State::Unresolved`.
* Fix comment style and grammar.
* Remove `InstId::PoisonedNameIndex`.
* Use "non-poisoned" instead of "non poisoned".

bricknerb committed:
Change import poison test to be successful as expected

bricknerb committed:
Ignore poisoned entries from imported from API

bricknerb committed:
Add missing library declaration

bricknerb committed:
Add a test for using a poisoned name without trying to define it

bricknerb committed:
Rephrase poisoned name declaration error

bricknerb committed:
Avoid trying to poison the same name twice

bricknerb committed:
Make sure LookupOrAdd() is not called with a poisoned InstId

bricknerb committed:
Add a TODO to improve the poisoned name diagnostic

bricknerb committed:
Fix DiagnosePoisonedName() param name to match definition

bricknerb committed:
Add support for poisoning in interface declarations

bricknerb committed:
Add support for poisoning in namespaces declarations

bricknerb committed:
Test declaraing data member using a poisoned name

bricknerb committed:
Support name poisoning in function declaration

bricknerb committed:
Add missing library in test

bricknerb committed:
Remove extra SemIR::

jonmeow commented on toolchain/sem_ir/name_scope.h:
```diff
@@ -178,7 +186,9 @@ class NameScopeStore {
+  //
   // These must never conflict.
+  //
   // inst_id must not be poisoned.
```
```suggestion
  // Adds a name that is required to exist in a name scope, such as `Self`.
  // The name must never conflict. inst_id must not be poisoned.
```

bricknerb replied:
Done.

jonmeow commented on toolchain/sem_ir/name_scope.h:
```diff
@@ -61,18 +63,23 @@ class NameScope : public Printable<NameScope> {
   // If the name is not found, adds the name using inst_id and access_kind and
   // returns false and an EntryId.
+  //
+  // inst_id must not be poisoned.
```
```suggestion
  // If the given name already exists, return true with the EntryId; the entry
  // might be poisoned. Otherwise, adds the name using inst_id and access_kind
  // and returns false with the new EntryId.
  //
  // This cannot be used to add poisoned entries; use AddPoison instead.
```

Suggesting a slightly different phrasing here, since I'm suggesting the paragraph edit too.

bricknerb replied:
Done.

jonmeow commented on toolchain/sem_ir/name_scope.h:
```diff
@@ -51,6 +52,7 @@ class NameScope : public Printable<NameScope> {
   // Searches for the given name and returns an EntryId if found or nullopt if
   // not.
+  //
   // If the name is found, it might be a poisoned name.
```
```suggestion
  // Searches for the given name and returns an EntryId if found or nullopt if
  // not. The returned entry may be poisoned.
```

I'm going to suggest using fewer paragraphs. The context and briefer read can often be helpful, and paragraphs are typically a few sentences anyways.

bricknerb replied:
Done.

jonmeow commented on toolchain/sem_ir/name_scope.h:
```diff
@@ -61,18 +63,23 @@ class NameScope : public Printable<NameScope> {
 
   // Adds a new name known to not exist.
+  //
   // Must not be poisoned.
```
```suggestion
  // Adds a new name known to not exist. Must not be poisoned.
```

bricknerb replied:
Done.

zygoloid committed:
Rename mutable accessor in `InstBlock` store. (#4659)

Mutating a block is a strange and rare operation and shouldn't have an
innocuous name like `Get`.

bricknerb committed:
Mark some //common, //toolchain/driver, //‎toolchain/install tests as small per 'Test execution time' warning (#4646)

These tests only take between 0.1s and 1.4s.

zygoloid committed:
Rename various `TryToCompleteType` functions to better describe what they do. (#4658)

As requested in review of #4652.

jonmeow committed:
Refactor single-unit checking out of check.cpp (#4649)

This is primarily moving code around, to try to create a logical split
of the code in check.cpp, makingthe API boundaries clearer.

There's one small, deliberate logic change around false returns from
`HandleParseNode`, where before there was a `CARBON_CHECK` instantiated
by the `#define` (per `NodeKind`), and now it's outside the `#define`
(done mainly because the message didn't keep up with the `Handle##Name`
-> `HandleParseNode` rename).

bricknerb committed:
Rephrase comments

bricknerb committed:
Clarify TODO for improving name poisoning diagnostic

jonmeow committed:
Update the vscode extension for publishing. (#4660)

A few initial fixes just so that publishing works.


https://marketplace.visualstudio.com/items?itemName=carbon-lang.carbon-vscode

zygoloid committed:
Switch from recommending a local workspace extension to recommending our published extension. (#4661)

geoffromer committed:
Proposal: Variadics (#2240)

Proposes a set of core features for declaring and implementing generic
variadic
functions.

A "pack expansion" is a syntactic unit beginning with `...`, which is a
kind of
compile-time loop over sequences called "packs". Packs are initialized
and
referred to using "pack bindings", which are marked with the `each`
keyword at
the point of declaration and the point of use.

The syntax and behavior of a pack expansion depends on its context, and
in some
cases by a keyword following the `...`:

- In a tuple literal expression (such as a function call argument list),
`...`
iteratively evaluates its operand expression, and treats the values as
    successive elements of the tuple.
- `...and` and `...or` iteratively evaluate a boolean expression,
combining
the values using `and` and `or`, and ending the loop early if the
underlying
    operator short-circuits.
-   In a statement context, `...` iteratively executes a statement.
- In a tuple literal pattern (such as a function parameter list), `...`
iteratively matches the elements of the scrutinee tuple. In conjunction
with
    pack bindings, this enables functions to take an arbitrary number of
    arguments.

---------

Co-authored-by: josh11b <josh11b@users.noreply.github.com>
Co-authored-by: Richard Smith <richard@metafoo.co.uk>
Co-authored-by: josh11b <15258583+josh11b@users.noreply.github.com>
Co-authored-by: Chandler Carruth <chandlerc@gmail.com>
Co-authored-by: Carbon Infra Bot <carbon-external-infra@google.com>

zygoloid committed:
Include a fully-qualified name when stringifying types. (#4657)

For example, format the `ImplicitAs` interface as `Core.ImplicitAs`
rather than simply `ImplicitAs`.

When importing an entity in a namespace, also import a declaration of
the enclosing namespace if necessary so that we can determine its name.

jonmeow commented on toolchain/check/context.cpp:
```diff
@@ -216,6 +216,17 @@ auto Context::DiagnoseDuplicateName(SemIRLoc dup_def, SemIRLoc prev_def)
+  // TODO: Improve the diagnostic to output the error where the name was
+  // poisoned and a note the name was later declared. See discussion in
+  // https://github.com/carbon-language/carbon-lang/pull/4654#discussion_r1876607172
+  CARBON_DIAGNOSTIC(
+      NameDeclUsedUnqualifiedBefore, Error,
+      "name previously used by unqualified name lookup and not found; "
+      "cannot later be declared");
+  emitter_->Build(loc, NameDeclUsedUnqualifiedBefore).Emit();
```
```suggestion
  CARBON_DIAGNOSTIC(NameUseBeforeDecl, Error, "name used before it was declared");
  CARBON_DIAGNOSTIC(NameUseBeforeDeclNote, Note, "declared here");
  unit_info.emitter.Emit(Parse::NodeId::Invalid, NameUseBeforeDecl)
      .Note(loc, NameUseBeforeDeclNote)
      .Emit();
```

bricknerb replied:
Thanks! Done.
I was not aware that it's ok (for now) to put an invalid location, and that's why I thought we won't be able to error if we don't use the location we have.

jonmeow commented on toolchain/check/context.cpp:
```diff
@@ -216,6 +216,17 @@ auto Context::DiagnoseDuplicateName(SemIRLoc dup_def, SemIRLoc prev_def)
 }
 
+auto Context::DiagnosePoisonedName(SemIRLoc loc) -> void {
+  // TODO: Improve the diagnostic to output the error where the name was
+  // poisoned and a note the name was later declared. See discussion in
```
```suggestion
  // TODO: Improve the diagnostic to replace NodeId::Invalid with the location where the name was poisoned. See discussion in
```

bricknerb replied:
Done.

jonmeow committed:
Make more use of llvm STLExtras (#4668)

This is essentially the result of looking at `.begin()` uses. We also
frequently do `std::shuffle`, but unfortunately STLExtras doesn't
provide a wrapper for that.

zygoloid committed:
Pacify CHECK failure on invalid code. (#4665)

Found by fuzzer.

---------

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

zygoloid committed:
Produce a note indicating where the specific was used from if monomorphization fails. (#4662)

Also fix a bug in `Context::GetClassType` that previously tried to
complete the class type before returning it. That's not correct --
`GetCompleteTypeImpl` is only appropriate for cases where the type can
trivially be completed and completing it can't fail -- and led to
infinite recursion with this change because we would call `GetClassType`
when producing a diagnostic if completing that class type failed.

zygoloid committed:
Provide a location for monomorphization failures resulting from `TryToCompleteType`. (#4670)

Almost all callers actually could never fail and nearly all of those
already `CHECK`-failed on failure. Add a new overload for that case, and
add a location parameter for the one remaining call.

bricknerb committed:
Change name poisoning error to error without a location for now and note to the place where the name is declared after it was poisoned.

jonmeow committed:
Add `Dump` functions to Check, Parse, and Lex (#4669)

- Provide `Check::Dump(context, arg)` and similar.
- gdb and lldb should do contextual lookup, and `call Dump(*this,
Lex::TokenIndex::Invalid)` has been tested with gdb.
- Since this is only for debug, keeps the functions fully separated from
code.
- Uses alwayslink to ensure objects are correctly linked, even though
there are no calls.
- `-Wno-missing-prototypes` is needed when we don't have forward
declarations.
- Code is not linked in opt builds, using `#ifndef NDEBUG`.
- This probably could be doing something in BUILD files with a
`select()`, but the `#ifndef` seemed easier.

This is based on #4620, but uses free functions instead of member
functions.

Co-authored-by: Dana Jansens <danakj@orodu.net>

---------

Co-authored-by: danakj <danakj@orodu.net>

bricknerb committed:
Allow defining .h files in tests without trying to compile them as Carbon files (#4667)

This would be used to test interop with C++.
#4666

zygoloid committed:
Lower global variables as global definitions, not global declarations. (#4674)

zygoloid committed:
Import support for array types. (#4675)

zygoloid committed:
Support stringifying tuple values. (#4664)

jonmeow committed:
Use a filename without a line number as a cue for autoupdate. (#4677)

This is so that diagnostics which lack a location get split file
clustering.

jonmeow committed:
Clean up missing library in test (#4678)

Noted in #4677

danakj committed:
Avoid printing enums as characters (#4676)

Given code like the following:
```
auto kind = ConversionTarget::Kind{0};
CARBON_CHECK(!loc_id.is_valid(), "hello {0} world", kind);
```

Currently we would print 'hello <the next line>', as the check string
would be treated as terminating at the '{0}', so it does not print the
rest of the string or a newline. This is because ConversionTarget::Kind
is an enum with underlying type `int8_t` which is a char, and
llvm::formatv does not look if the type is an enum and treat is
specially. So it prints it as a char rather than a number, which in this
case is a nul terminator.

With this change, the '{0}' value will be converted to a larger integer
before being passed through to llvm::formatv so that char-sized enums
will print as a number, and the result is that we will print 'hello 0
world\n' as the developer intended.

danakj committed:
Add a --remote switch to new_proposal.py (#4681)

If the user's fork is not named 'origin' then the script will fail and
needs to know the user's remote name.

Fixes #1899

jonmeow commented:
Can you take a look at the test failures?

bricknerb commented:
> Can you take a look at the test failures?

Sure, I couldn't reproduce this locally and couldn't find a diff between what I run locally and the failure.
I obviously do something wrong here.
Do you have any pointers how can I reproduce this?

zygoloid commented:
It's probably only failing after merging with trunk.

jonmeow committed:
Fix mem usage tracking of semir (#4684)

The call got misplaced during refactoring.

zygoloid committed:
Fix lowering of array indexing with an int literal. (#4686)

Such indexing operations are created by array initialization. Since we
switched integer literals to be of type IntLiteral we've been attempting
to index arrays with the (empty) representation of an IntLiteral rather
than with an actual integer value.

bricknerb committed:
Update name poisoning test following merge with trunk

bricknerb committed:
Merge branch 'trunk' into poison4

bricknerb commented:
> It's probably only failing after merging with trunk.

Thanks!
This makes sense, just wasn't aware this is what it does.
I think I've fixed it, but I can probably not rerun it myself.

jonmeow committed:
Merge trunk

jonmeow commented:
I see this failed on another merge conflict. I've merged to try to get it in now. The contributor access should help avoid this repeating, though; I encourage resolving conflicts & merging after approval, now that you can. :)

jonmeow commented:
And maybe I just missed that I hadn't approved...

bricknerb commented:
Thanks for the merge Jon!
Indeed I was waiting for an approval. :)
